File: app/frontend/src/api/inventory.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import apiClient from './client'

// ─── Types ────────────────────────────────────────────────────────────────────

export type ItemType = 'server' | 'database' | 'email_account' | 'cloud_account' | 'generic'
export type GroupType = 'replication' | 'cluster' | 'ha' | 'load_balanced' | 'related' | 'other'

export interface InventoryGroup {
  id: string
  name: string
  description?: string
  group_type: GroupType
  color: string
  item_count: number
  created_at: string
  updated_at: string
}

export interface InventoryGroupCreate {
  name: string
  description?: string
  group_type?: GroupType
  color?: string
}

export interface InventoryGroupUpdate {
  name?: string
  description?: string
  group_type?: GroupType
  color?: string
}

export interface InventoryGroupSummary {
  id: string
  name: string
  group_type: GroupType
  color: string
}

export interface InventoryItem {
  id: string
  item_type: ItemType
  display_name: string
  description?: string
  notes?: string
  owner?: string
  tags: string[]
  is_active: boolean

  // Server / shared
  hostname?: string
  ip_address?: string
  port?: number
  username?: string
  has_password: boolean
  has_ssh_key: boolean
  operating_system?: string

  // Database
  database_name?: string
  database_type?: string

  // Email
  email_address?: string
  smtp_host?: string
  smtp_port?: number
  imap_host?: string
  imap_port?: number

  // Cloud
  provider?: string
  account_id?: string
  has_access_key: boolean
  region?: string

  // Generic
  url?: string

  // Group
  group_id?: string
  group?: InventoryGroupSummary

  created_by?: string
  updated_by?: string
  created_at: string
  updated_at: string
}

export interface InventoryItemCreate {
  item_type: ItemType
  display_name: string
  description?: string
  notes?: string
  owner?: string
  tags?: string[]
  is_active?: boolean
  hostname?: string
  ip_address?: string
  port?: number
  username?: string
  password?: string
  ssh_key?: string
  operating_system?: string
  database_name?: string
  database_type?: string
  email_address?: string
  smtp_host?: string
  smtp_port?: number
  imap_host?: string
  imap_port?: number
  provider?: string
  account_id?: string
  access_key_id?: string
  secret_access_key?: string
  region?: string
  url?: string
  group_id?: string | null
}

export type InventoryItemUpdate = Partial<Omit<InventoryItemCreate, 'item_type'>> & { group_id?: string | null }

export interface InventorySchedule {
  id: string
  name: string
  frequency: 'daily' | 'weekly' | 'monthly'
  day_of_week?: number
  day_of_month?: number
  hour: number
  recipient_emails: string[]
  is_active: boolean
  created_by?: string
  last_run_at?: string
  next_run_at?: string
  created_at: string
  updated_at: string
}

export interface InventoryScheduleCreate {
  name: string
  frequency: 'daily' | 'weekly' | 'monthly'
  day_of_week?: number
  day_of_month?: number
  hour?: number
  recipient_emails: string[]
  is_active?: boolean
}

// ─── Query keys ──────────────────────────────────────────────────────────────

const inventoryKeys = {
  all: ['inventory'] as const,
  items: (params?: object) => [...inventoryKeys.all, 'items', params] as const,
  item: (id: string) => [...inventoryKeys.all, 'item', id] as const,
  groups: () => [...inventoryKeys.all, 'groups'] as const,
  schedules: () => [...inventoryKeys.all, 'schedules'] as const,
}

// ─── Item Hooks ───────────────────────────────────────────────────────────────

export interface InventoryListParams {
  item_type?: string
  search?: string
  tags?: string
  is_active?: boolean
  group_id?: string
  skip?: number
  limit?: number
}

export function useInventoryItems(params?: InventoryListParams) {
  return useQuery({
    queryKey: inventoryKeys.items(params),
    queryFn: async () => {
      const searchParams = new URLSearchParams()
      if (params?.item_type) searchParams.set('item_type', params.item_type)
      if (params?.search) searchParams.set('search', params.search)
      if (params?.tags) searchParams.set('tags', params.tags)
      if (params?.is_active !== undefined) searchParams.set('is_active', String(params.is_active))
      if (params?.group_id) searchParams.set('group_id', params.group_id)
      if (params?.skip !== undefined) searchParams.set('skip', String(params.skip))
      if (params?.limit !== undefined) searchParams.set('limit', String(params.limit))
      const qs = searchParams.toString()
      const res = await apiClient.get<InventoryItem[]>(`/inventory/items${qs ? `?${qs}` : ''}`)
      return res.data
    },
  })
}

export function useCreateInventoryItem() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: InventoryItemCreate) =>
      apiClient.post<InventoryItem>('/inventory/items', data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: inventoryKeys.all }),
  })
}

export function useUpdateInventoryItem() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, data }: { id: string; data: InventoryItemUpdate }) =>
      apiClient.patch<InventoryItem>(`/inventory/items/${id}`, data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: inventoryKeys.all }),
  })
}

export function useDeleteInventoryItem() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) =>
      apiClient.delete(`/inventory/items/${id}`).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: inventoryKeys.all }),
  })
}

export function useRevealField() {
  return useMutation({
    mutationFn: ({ id, field }: { id: string; field: string }) =>
      apiClient
        .post<{ field: string; value: string }>(`/inventory/items/${id}/reveal`, { field })
        .then((r) => r.data),
  })
}

// ─── Export ───────────────────────────────────────────────────────────────────

export async function exportInventory(format: 'excel' | 'csv', item_type?: string, scope: 'all' | 'visible' = 'all') {
  const params = new URLSearchParams({ format, scope })
  if (item_type) params.set('item_type', item_type)

  const res = await apiClient.get(`/inventory/export?${params.toString()}`, {
    responseType: 'blob',
  })

  const ext = format === 'csv' ? 'csv' : 'xlsx'
  const contentType =
    format === 'csv'
      ? 'text/csv'
      : 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'

  const url = URL.createObjectURL(new Blob([res.data], { type: contentType }))
  const a = document.createElement('a')
  a.href = url
  a.download = `inventory.${ext}`
  a.click()
  URL.revokeObjectURL(url)
}

// ─── Group Hooks ─────────────────────────────────────────────────────────────

export function useInventoryGroups() {
  return useQuery({
    queryKey: inventoryKeys.groups(),
    queryFn: async () => {
      const res = await apiClient.get<InventoryGroup[]>('/inventory/groups')
      return res.data
    },
  })
}

export function useCreateInventoryGroup() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: InventoryGroupCreate) =>
      apiClient.post<InventoryGroup>('/inventory/groups', data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: inventoryKeys.all }),
  })
}

export function useUpdateInventoryGroup() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, data }: { id: string; data: InventoryGroupUpdate }) =>
      apiClient.patch<InventoryGroup>(`/inventory/groups/${id}`, data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: inventoryKeys.all }),
  })
}

export function useDeleteInventoryGroup() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) =>
      apiClient.delete(`/inventory/groups/${id}`).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: inventoryKeys.all }),
  })
}

export function useAssignItemsToGroup() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ groupId, itemIds }: { groupId: string | null; itemIds: string[] }) => {
      if (groupId === null) {
        return apiClient.post<{ updated: number }>('/inventory/groups/unassign', { item_ids: itemIds }).then((r) => r.data)
      }
      return apiClient.post<{ updated: number }>(`/inventory/groups/${groupId}/assign`, { item_ids: itemIds }).then((r) => r.data)
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: inventoryKeys.all }),
  })
}

// ─── Schedule Hooks ───────────────────────────────────────────────────────────

export function useInventorySchedules() {
  return useQuery({
    queryKey: inventoryKeys.schedules(),
    queryFn: async () => {
      const res = await apiClient.get<InventorySchedule[]>('/inventory/schedules')
      return res.data
    },
  })
}

export function useCreateInventorySchedule() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: InventoryScheduleCreate) =>
      apiClient.post<InventorySchedule>('/inventory/schedules', data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: inventoryKeys.schedules() }),
  })
}

export function useUpdateInventorySchedule() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({
      id,
      data,
    }: {
      id: string
      data: Partial<InventoryScheduleCreate>
    }) =>
      apiClient
        .patch<InventorySchedule>(`/inventory/schedules/${id}`, data)
        .then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: inventoryKeys.schedules() }),
  })
}

export function useDeleteInventorySchedule() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) =>
      apiClient.delete(`/inventory/schedules/${id}`).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: inventoryKeys.schedules() }),
  })
}

export function useSendInventoryScheduleNow() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) =>
      apiClient.post<{ sent: number }>(`/inventory/schedules/${id}/send-now`).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: inventoryKeys.schedules() }),
  })
}

File: app/backend/app/schemas/inventory.py
import ipaddress
import uuid
from datetime import datetime
from typing import Literal, Optional
from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


VALID_ITEM_TYPES = ["server", "database", "email_account", "cloud_account", "generic"]
VALID_FREQUENCIES = ["daily", "weekly", "monthly"]
VALID_GROUP_TYPES = ["replication", "cluster", "ha", "load_balanced", "related", "other"]


# ── Group schemas ─────────────────────────────────────────────────────────────

class InventoryGroupCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=2000)
    group_type: str = Field("related")
    color: str = Field("#6366f1")


class InventoryGroupUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=2000)
    group_type: Optional[str] = None
    color: Optional[str] = None


class InventoryGroupSummary(BaseModel):
    id: uuid.UUID
    name: str
    group_type: str
    color: str
    model_config = ConfigDict(from_attributes=True)


class InventoryGroupResponse(BaseModel):
    id: uuid.UUID
    name: str
    description: Optional[str] = None
    group_type: str
    color: str
    item_count: int = 0
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


class InventoryGroupAssign(BaseModel):
    item_ids: list[uuid.UUID]


# ── Item schemas ──────────────────────────────────────────────────────────────

class InventoryItemCreate(BaseModel):
    item_type: Literal["server", "database", "email_account", "cloud_account", "generic"]
    display_name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=5000)
    notes: Optional[str] = Field(None, max_length=5000)
    owner: Optional[str] = Field(None, max_length=255)
    tags: list[str] = Field(default_factory=list)
    is_active: bool = True

    # Server / DB / shared
    hostname: Optional[str] = Field(None, max_length=255)
    ip_address: Optional[str] = Field(None, max_length=45)
    port: Optional[int] = Field(None, ge=1, le=65535)
    username: Optional[str] = Field(None, max_length=255)
    password: Optional[str] = None          # plaintext → encrypted in service
    ssh_key: Optional[str] = None           # plaintext → encrypted in service
    operating_system: Optional[str] = Field(None, max_length=100)

    # Database-specific
    database_name: Optional[str] = Field(None, max_length=255)
    database_type: Optional[str] = Field(None, max_length=50)

    # Email account
    email_address: Optional[str] = Field(None, max_length=255)
    smtp_host: Optional[str] = Field(None, max_length=255)
    smtp_port: Optional[int] = Field(None, ge=1, le=65535)
    imap_host: Optional[str] = Field(None, max_length=255)
    imap_port: Optional[int] = Field(None, ge=1, le=65535)

    # Cloud account
    provider: Optional[str] = Field(None, max_length=50)
    account_id: Optional[str] = Field(None, max_length=255)
    access_key_id: Optional[str] = None    # plaintext → encrypted
    secret_access_key: Optional[str] = None  # plaintext → encrypted
    region: Optional[str] = Field(None, max_length=100)

    # Generic
    url: Optional[str] = Field(None, max_length=1000)

    # Group assignment
    group_id: Optional[uuid.UUID] = None

    @field_validator("ip_address")
    @classmethod
    def validate_ip(cls, v: Optional[str]) -> Optional[str]:
        if v is None or v == "":
            return v
        try:
            ipaddress.ip_address(v)
        except ValueError:
            raise ValueError("Geçersiz IP adresi.")
        return v

    @field_validator("url")
    @classmethod
    def validate_url(cls, v: Optional[str]) -> Optional[str]:
        if v is None or v == "":
            return v
        if not (v.startswith("http://") or v.startswith("https://") or "://" in v):
            raise ValueError("Geçersiz URL.")
        return v


class InventoryItemUpdate(BaseModel):
    display_name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=5000)
    notes: Optional[str] = Field(None, max_length=5000)
    owner: Optional[str] = Field(None, max_length=255)
    tags: Optional[list[str]] = None
    is_active: Optional[bool] = None

    hostname: Optional[str] = Field(None, max_length=255)
    ip_address: Optional[str] = Field(None, max_length=45)
    port: Optional[int] = Field(None, ge=1, le=65535)
    username: Optional[str] = Field(None, max_length=255)
    password: Optional[str] = None
    ssh_key: Optional[str] = None
    operating_system: Optional[str] = Field(None, max_length=100)

    database_name: Optional[str] = Field(None, max_length=255)
    database_type: Optional[str] = Field(None, max_length=50)

    email_address: Optional[str] = Field(None, max_length=255)
    smtp_host: Optional[str] = Field(None, max_length=255)
    smtp_port: Optional[int] = Field(None, ge=1, le=65535)
    imap_host: Optional[str] = Field(None, max_length=255)
    imap_port: Optional[int] = Field(None, ge=1, le=65535)

    provider: Optional[str] = Field(None, max_length=50)
    account_id: Optional[str] = Field(None, max_length=255)
    access_key_id: Optional[str] = None
    secret_access_key: Optional[str] = None
    region: Optional[str] = Field(None, max_length=100)

    url: Optional[str] = Field(None, max_length=1000)

    # Group assignment (None = remove from group)
    group_id: Optional[uuid.UUID] = None

    @field_validator("ip_address")
    @classmethod
    def validate_ip(cls, v: Optional[str]) -> Optional[str]:
        if v is None or v == "":
            return v
        try:
            ipaddress.ip_address(v)
        except ValueError:
            raise ValueError("Geçersiz IP adresi.")
        return v


class InventoryItemResponse(BaseModel):
    id: uuid.UUID
    item_type: str
    display_name: str
    description: Optional[str] = None
    notes: Optional[str] = None
    owner: Optional[str] = None
    tags: list[str] = []
    is_active: bool

    # Server / shared
    hostname: Optional[str] = None
    ip_address: Optional[str] = None
    port: Optional[int] = None
    username: Optional[str] = None
    has_password: bool = False
    has_ssh_key: bool = False
    operating_system: Optional[str] = None

    # Database
    database_name: Optional[str] = None
    database_type: Optional[str] = None

    # Email
    email_address: Optional[str] = None
    smtp_host: Optional[str] = None
    smtp_port: Optional[int] = None
    imap_host: Optional[str] = None
    imap_port: Optional[int] = None

    # Cloud
    provider: Optional[str] = None
    account_id: Optional[str] = None
    has_access_key: bool = False
    region: Optional[str] = None

    # Generic
    url: Optional[str] = None

    # Group
    group_id: Optional[uuid.UUID] = None
    group: Optional[InventoryGroupSummary] = None

    created_by: Optional[uuid.UUID] = None
    updated_by: Optional[uuid.UUID] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class InventoryRevealRequest(BaseModel):
    field: Literal["password", "ssh_key", "access_key_id", "secret_access_key"]


class InventoryRevealResponse(BaseModel):
    field: str
    value: str


# ── Schedule schemas ──────────────────────────────────────────────────────────

class InventoryScheduleCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    frequency: Literal["daily", "weekly", "monthly"] = "weekly"
    day_of_week: Optional[int] = Field(None, ge=0, le=6)
    day_of_month: Optional[int] = Field(None, ge=1, le=31)
    hour: int = Field(8, ge=0, le=23)
    recipient_emails: list[str] = Field(..., min_length=1)
    is_active: bool = True
    teams_webhook_id: Optional[uuid.UUID] = None

    @model_validator(mode="after")
    def validate_schedule(self) -> "InventoryScheduleCreate":
        if self.frequency == "weekly" and self.day_of_week is None:
            raise ValueError("Haftalık sıklık için haftanın günü belirtilmelidir.")
        if self.frequency == "monthly" and self.day_of_month is None:
            raise ValueError("Aylık sıklık için ayın günü belirtilmelidir.")
        return self


class InventoryScheduleUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    frequency: Optional[Literal["daily", "weekly", "monthly"]] = None
    day_of_week: Optional[int] = Field(None, ge=0, le=6)
    day_of_month: Optional[int] = Field(None, ge=1, le=31)
    hour: Optional[int] = Field(None, ge=0, le=23)
    recipient_emails: Optional[list[str]] = None
    is_active: Optional[bool] = None
    teams_webhook_id: Optional[uuid.UUID] = None


class InventoryScheduleResponse(BaseModel):
    id: uuid.UUID
    name: str
    frequency: str
    day_of_week: Optional[int] = None
    day_of_month: Optional[int] = None
    hour: int
    recipient_emails: list[str]
    is_active: bool
    teams_webhook_id: Optional[uuid.UUID] = None
    created_by: Optional[uuid.UUID] = None
    last_run_at: Optional[datetime] = None
    next_run_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
